sound/soc/sof/ipc4-loader.c

```diff
@@ -167,13 +170,86 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
 	return payload_offset;
 }
 
+static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
+					 const guid_t *uuid)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_ipc4_fw_library *fw_lib;
+	const char *fw_filename;
+	size_t payload_offset;
+	u32 lib_id;
+	int ret, i;
+
+	if (ipc4_data->max_libs_count == 1) {
+		dev_err(sdev->dev, "Loaded basefw does not support external libraries\n");
+		return -EOPNOTSUPP;
+	}
+
+	if (!sdev->pdata->fw_lib_prefix) {
+		dev_err(sdev->dev,
+			"Library loading is not supported due to not set library path\n");
+		return -EINVAL;
+	}
+
+	if (!ipc4_data->load_library) {
+		dev_err(sdev->dev, "Library loading is not supported on this platform\n");
+		return -EOPNOTSUPP;
+	}
+
+	fw_lib = devm_kzalloc(sdev->dev, sizeof(*fw_lib), GFP_KERNEL);
+	if (!fw_lib)
+		return -ENOMEM;
+
+	fw_filename = kasprintf(GFP_KERNEL, "%s/%pUL.bin",
+				sdev->pdata->fw_lib_prefix, uuid);
+	if (!fw_filename)
+		return -ENOMEM;
+
+	ret = request_firmware(&fw_lib->sof_fw.fw, fw_filename, sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "Library file '%s' is missing\n", fw_filename);
+		goto free;
+	} else {
+		dev_dbg(sdev->dev, "Library file '%s' loaded\n", fw_filename);
+	}
+
+	payload_offset = sof_ipc4_fw_parse_ext_man(sdev, fw_lib);
+	if (payload_offset > 0) {
+		fw_lib->sof_fw.payload_offset = payload_offset;
+		ret = xa_alloc(&ipc4_data->fw_lib_xa, &lib_id, fw_lib,
+			       XA_LIMIT(1, ipc4_data->max_libs_count - 1), GFP_KERNEL);
+		if (ret)
+			goto release;
+
+		fw_lib->id = lib_id;
+	}
+
+	/* Fix up the module ID numbers within the library */
+	for (i = 0; i < fw_lib->num_modules; i++)
+		fw_lib->modules[i].man4_module_entry.id += (SOF_IPC4_MODULE_ID_STRIDE * lib_id);
+
+	kfree(fw_filename);
+
+	return ipc4_data->load_library(sdev, fw_lib, false);
```

Comment:
Note: we will keep the library in the kernel list even if loading fails. But this will only happen during tplg parsing at the moment and if that fails, the probe will also fail and thus it does not really matter since the SOF stack is not initialized to be usable.

Other use cases will need more planning to cover all corner cases.

Reply:
I will not leave it like this. Let me find a solution which has no holes and TODOs.